rmw_fastrtps_shared_cpp/src/rmw_get_topic_info.cpp

```diff
@@ -0,0 +1,306 @@
+  rmw_topic_info_array_t * participants_info)
+{
+  if (!identifier) {
+    RMW_SET_ERROR_MSG("null implementation identifier provided");
+    return RMW_RET_ERROR;
+  }
+
+  if (!topic_name) {
+    RMW_SET_ERROR_MSG("null topic_name provided");
+    return RMW_RET_ERROR;
+  }
+
+  if (!allocator) {
+    RMW_SET_ERROR_MSG("null allocator provided");
+    return RMW_RET_ERROR;
+  }
+
+  if (!node) {
+    RMW_SET_ERROR_MSG("null node handle");
+    return RMW_RET_ERROR;
+  }
+
+  // Get participant pointer from node
+  if (node->implementation_identifier != identifier) {
+    RMW_SET_ERROR_MSG("node handle not from this implementation");
+    return RMW_RET_ERROR;
+  }
+
+  if (!participants_info) {
+    RMW_SET_ERROR_MSG("null participants_info provided");
+    return RMW_RET_ERROR;
+  }
+
+  return RMW_RET_OK;
+}
+
+std::vector<std::string>
+_get_topic_fqdns(const char * topic_name, bool no_mangle)
+{
+  std::vector<std::string> topic_fqdns;
+  topic_fqdns.push_back(topic_name);
+  if (!no_mangle) {
+    auto ros_prefixes = _get_all_ros_prefixes();
+    // Build the list of all possible topic FQDN
+    std::for_each(ros_prefixes.begin(), ros_prefixes.end(),
+      [&topic_fqdns, &topic_name](const std::string & prefix) {
+        topic_fqdns.push_back(prefix + topic_name);
+      });
+  }
+  return topic_fqdns;
+}
+
+void
+_handle_topic_info_fini(
+  rmw_topic_info_t * topic_info,
+  rcutils_allocator_t * allocator)
+{
+  bool had_error = rmw_error_is_set();
+  std::string error_string;
+  if (had_error) {
+    error_string = rmw_get_error_string().str;
+  }
+  rmw_reset_error();
+
+  rmw_ret_t ret = rmw_topic_info_fini(topic_info, allocator);
+  if (ret != RMW_RET_OK) {
+    RCUTILS_LOG_ERROR_NAMED(
+      "rmw_fastrtps_cpp",
+      "rmw_topic_info_fini failed: %s",
+      rmw_get_error_string().str);
+    rmw_reset_error();
+  }
+
+  if (had_error) {
+    RMW_SET_ERROR_MSG(error_string.c_str());
+  }
+}
+
+rmw_ret_t
+_set_rmw_topic_info(
+  rmw_topic_info_t * topic_info,
+  const GUID_t & participant_guid,
+  const char * node_name,
+  const char * node_namespace,
+  TopicData topic_data,
+  ::ParticipantListener * slave_target,
+  rcutils_allocator_t * allocator)
+{
+  static_assert(
+    sizeof(eprosima::fastrtps::rtps::GUID_t) <= RMW_GID_STORAGE_SIZE,
+    "RMW_GID_STORAGE_SIZE insufficient to store the rmw_fastrtps_cpp GID implementation."
+  );
+  uint8_t rmw_gid[RMW_GID_STORAGE_SIZE];
+  memset(&rmw_gid, 0, RMW_GID_STORAGE_SIZE);
+  memcpy(&rmw_gid, &topic_data.entity_guid, sizeof(eprosima::fastrtps::rtps::GUID_t));
+  rmw_ret_t ret = rmw_topic_info_set_gid(topic_info, rmw_gid,
+      sizeof(eprosima::fastrtps::rtps::GUID_t));
+  if (ret != RMW_RET_OK) {
+    return ret;
+  }
+  // set qos profile
+  ret = rmw_topic_info_set_qos_profile(topic_info, &topic_data.qos_profile);
+  if (ret != RMW_RET_OK) {
+    return ret;
+  }
+  // set topic type
+  ret = rmw_topic_info_set_topic_type(topic_info, topic_data.topic_type.c_str(), allocator);
```

Comment:
Please, demangle the topic type here (see my other comment https://github.com/ros2/ros2cli/pull/385#issuecomment-567019680).

See https://github.com/ros2/rmw_fastrtps/blob/8f6a13c5995b13a8bd9a0d147851ab5754bdf104/rmw_fastrtps_shared_cpp/src/demangle.hpp#L26.

Reply:
Done.

Reply:
Extra comment about this, that I didn't think before:
In case the user passes `no_mangle=true`, I think that it makes sense to in that case to show the unmangled type name (i.e.: call `_demangle_if_ros_type` only when `no_mangle` is `true`).

I don't see the value of having two flags in the function signature, i.e.: one for specifying if the passed topic name will be mangled with the ros conventions, and one for indicating if the discovered type should be demangled or not. But that's also an option.
I don't care much about what of the two options is followed, though the first one is simpler.

Please also, update the documentation in the `rmw` PR, so this detail is clearer.

Reply:
I think need some context here, @jaisontj and @prajakta-gokhale. If I use `no_mangle=true` in the `rcl` and `rclpy` tests, there are failures, like the number of found publishers and subscriptions is 0.

Reply:
Sorry, I will try to give some context.
The no_mangle/no_demangle options are intended for applications that mix ROS 2 with native DDS. The idea is to avoid the name mangling that ROS does when creating topic names and type names.
For example:
/my_topic is translated in something like `rt/my_topic`
/my_service creates two dds topics (I don't remember well these mangling rules).

When you use `no_mangle=true`, it stops working because the topic name is not found.
It's not a good idea to add tests with `no_mangle=true` in rcl, as mangling rules might be middleware dependent.

Hope this helps to clarify.

Reply:
Okay, thanks for clarifying! I've made demangling conditional based on `no_mangle == true`. The `rcl` and `rclpy` unit tests will continue to test against non-demangled type names.